src/commonMain/kotlin/fr/acinq/lightning/channel/states/Channel.kt

```diff
@@ -161,7 +161,75 @@ sealed class ChannelState {
+        logger.error(t) { "error on command ${cmd::class.simpleName}" }
+
+        fun abort(channelId: ByteVector32?): Pair<ChannelState, List<ChannelAction>> {
+            val error = channelId?.let { Error(it, t.message) }
+            val actions = error?.let { listOf(ChannelAction.Message.Send(it)) } ?: emptyList()
+            return Pair(Aborted, actions)
+        }
+
+        fun forceClose(state: ChannelStateWithCommitments): Pair<ChannelState, List<ChannelAction>> {
+            val error = Error(state.channelId, t.message)
+            return when {
+                state.commitments.nothingAtStake() -> Pair(Aborted, listOf(ChannelAction.Message.Send(error)))
+                else -> state.run { spendLocalCurrent().run { copy(second = second + ChannelAction.Message.Send(error)) } }
+            }
+        }
+
+        return when(val state = this@ChannelState) {
+            is WaitForInit -> abort(null)
+            is WaitForOpenChannel -> abort(state.temporaryChannelId)
+            is WaitForAcceptChannel -> abort(state.temporaryChannelId)
+            is WaitForFundingCreated -> abort(state.channelId)
+            is WaitForFundingSigned -> abort(state.channelId)
+            is WaitForFundingConfirmed -> forceClose(state)
+            is WaitForChannelReady -> forceClose(state)
+            is Normal -> forceClose(state)
+            is ShuttingDown -> forceClose(state)
+            is Negotiating -> {
+                val error = Error(state.channelId, t.message)
+                when {
+                    state.commitments.nothingAtStake() -> Pair(Aborted, listOf(ChannelAction.Message.Send(error)))
+                    state.bestUnpublishedClosingTx != null -> {
+                        // if we were in the process of closing and already received a closing sig from the counterparty, it's always better to use that
+                        val nextState = Closing(
+                            state.commitments,
+                            waitingSinceBlock = currentBlockHeight.toLong(),
+                            mutualCloseProposed = state.closingTxProposed.flatten().map { it.unsignedTx } + listOf(state.bestUnpublishedClosingTx),
+                            mutualClosePublished = listOf(state.bestUnpublishedClosingTx)
+                        )
+                        val actions = listOf(
+                            ChannelAction.Storage.StoreState(nextState),
+                            ChannelAction.Blockchain.PublishTx(state.bestUnpublishedClosingTx),
+                            ChannelAction.Blockchain.SendWatch(WatchConfirmed(state.channelId, state.bestUnpublishedClosingTx.tx, staticParams.nodeParams.minDepthBlocks.toLong(), BITCOIN_TX_CONFIRMED(state.bestUnpublishedClosingTx.tx)))
+                        )
+                        Pair(nextState, actions)
+                    }
+                    else -> state.run { spendLocalCurrent().run { copy(second = second + ChannelAction.Message.Send(error)) } }
```

Comment:
This could be factored as:

```suggestion
                    state.bestUnpublishedClosingTx != null -> {
                        // if we were in the process of closing and already received a closing sig from the counterparty, it's always better to use that
                        val nextState = Closing(
                            state.commitments,
                            waitingSinceBlock = currentBlockHeight.toLong(),
                            mutualCloseProposed = state.closingTxProposed.flatten().map { it.unsignedTx } + listOf(state.bestUnpublishedClosingTx),
                            mutualClosePublished = listOf(state.bestUnpublishedClosingTx)
                        )
                        val actions = listOf(
                            ChannelAction.Storage.StoreState(nextState),
                            ChannelAction.Blockchain.PublishTx(state.bestUnpublishedClosingTx),
                            ChannelAction.Blockchain.SendWatch(WatchConfirmed(state.channelId, state.bestUnpublishedClosingTx.tx, staticParams.nodeParams.minDepthBlocks.toLong(), BITCOIN_TX_CONFIRMED(state.bestUnpublishedClosingTx.tx)))
                        )
                        Pair(nextState, actions)
                    }
                    else -> forceClose(state)
```

Reply:
Nice! Done in [1946282](https://github.com/ACINQ/lightning-kmp/pull/479/commits/19462825456fb8133bfb79562f902c2546328de2).